### PR TITLE
fix: add unprivileged user to the Dockerfiles

### PR DESCRIPTION
**Pull request by levivannoort**
Close: #4895 

## PR Details

Added a standard setup for setting a non-root (unpriviliged) user to the Dockerfiles. This is a done to fix several of the security alerts raised within github by the trivy vulnerability scanning.

Security scanning for this pr shows that we got rid of the alerts (except for the ones in to-do): https://github.com/amplication/amplication/security/code-scanning?query=is%3Aopen+pr%3A5065

## PR Checklist

- [ ] Tests for the changes have been added
- [ ] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

### CODE REVIEW

File: packages/amplication-data-service-generator-runner/Dockerfile [hunk 1]
```diff
@@ -1,6 +1,19 @@
-FROM node:19.1-alpine AS base
+FROM node:18.13.0-slim AS base
```
Comment (overbit) [inline packages/amplication-data-service-generator-runner/Dockerfile:1]: `node:18.13.0-slim` is debian based and it won't have `apk` package manager. 
We either need to use the alpine version or fix line 8. 

My suggestion is to revert to an alpine based image and update the user configuration https://wiki.alpinelinux.org/wiki/Setting_up_a_new_user

File: packages/amplication-server/Dockerfile [hunk 2]
```diff
@@ -1,18 +1,25 @@
...
+COPY ./dist/packages/amplication-${component} .
 
+RUN groupadd --gid ${gid} ${user}
+RUN useradd --uid ${uid} --gid ${gid} -m ${user}
+RUN chown -R ${uid}:${gid} /workspace/packages/${user}
+
+USER ${user}
 
 ENV PORT=3333
 EXPOSE $PORT
```
Comment (overbit) [inline packages/amplication-server/Dockerfile:25]: amplication-server container doesn't start due to permissions.

```sh
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] MailModule dependencies initialized +0ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] BillingModule dependencies initialized +0ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] TopicModule dependencies initialized +1ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] CommitModule dependencies initialized +0ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] SubscriptionModule dependencies initialized +0ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] PluginInstallationModule dependencies initialized +0ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] GraphQLModule dependencies initialized +0ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] ResourceModule dependencies initialized +1ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] GitProviderModule dependencies initialized +0ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] AccountModule dependencies initialized +0ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] BuildModule dependencies initialized +1ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] ProjectModule dependencies initialized +0ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] WorkspaceModule dependencies initialized +0ms
[Nest] 1  - 01/25/2023, 11:53:41 AM     LOG [InstanceLoader] AuthModule dependencies initialized +0ms
[Nest] 1  - 01/25/2023, 11:53:41 AM   ERROR [ServerKafka] ERROR [BrokerPool] Failed to connect to seed broker, trying another broker from the list: Failed to connect: Port should be >= 0 and < 65536. Received type number (NaN). {"timestamp":"2023-01-25T11:53:41.464Z","logger":"kafkajs","retryCount":0,"retryTime":303}
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

[Error: EACCES: permission denied, open '../.server-id'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'open',
  path: '../.server-id'
}

Node.js v18.13.0
```

To test it locally I suggest to run `npx nx docker:build amplication-server` and `docker run --env-file packages/amplication-server/.env amplication/server`

**Review by overbit — CHANGES_REQUESTED** [review]

File: packages/amplication-server/Dockerfile (hunk 2 above)
Comment (overbit) [reply packages/amplication-server/Dockerfile:25]: this potentially affects other container images too but I have not check them yet

Comment (levivannoort) [reply packages/amplication-server/Dockerfile:25]: the problem here was the change ownership command that was a directory to far down, it was changed to:
```
RUN chown -R ${uid}:${gid} /workspace/packages/
```

**Comment by levivannoort** [comment]
**Note:** checkout the data-service-generator-runner volume mount (local and kubernetes) and make sure that this can work in conjunction with the non-priviliged user.

**Comment by overbit** [comment]
@levivannoort  keep in mind that we have to give R/W permissions to volumes mounted for these `non-root` user. 
I.e. for container like DSG-runner and build-manager that share a volume for dsg jobs and aritfacts `/efs`

**Comment by levivannoort** [comment]
When looking at the generated app, the db migrate didnt run properly so this should be tested for the server image.

**Comment by mulygottlieb** [comment]
@levivannoort This PR is open as a draft since February.  Please decide what to do with it - either complete it and change to non-draft for approval, or delete it.

**Comment by levivannoort** [comment]
This was blocked due to work for the audit, as the behaviour of the changes was quite impactful, I'd still be up for going ahead and movign forward with these changes as they should circumvent, 10-12 `high` priority security/vulnerability warnings in our scanning solution:

https://github.com/amplication/amplication/security/code-scanning

As Daniele mentioned before this also requires some work on the [infrastructure side](https://github.com/amplication/amplication-manifests/pull/50/files).

**Comment by levivannoort** [comment]
@mulygottlieb this does require quite a bit of effort (since we work with mounted volumes that are attached after at launch time - which need to have the right permissions et cetera) to get right on the infrastructure side of things most likely, so the question is if we would want to invest that time right now, as mentioned above it does fix quite a bit of the attack vector mentioned in the security 'alerts'.

I could spend half a day this friday; merge it, troubleshooting for sometime, revert if it takes too long and leave the branch open for another time. Your call!

**Comment by mulygottlieb** [comment]
@levivannoort I accept your suggestion - let's spend half a day on this (not urgent for this Friday - your call when you have this half-day between other tasks). If you can get it resolved - great. If not - revert and we will pick it up on another time (but perhaps in that case we need to remove the open PR and just leave the branch?)

**Comment by levivannoort** [comment]
@mulygottlieb closed for now as I am not able to get to it.

**Pull request closed without merging**